node/src/components/gossiper/gossip_table.rs

```diff
@@ -103,17 +103,53 @@ impl State {
     }
 }
 
+#[derive(Debug)]
+struct Timeouts<T> {
+    values: Vec<(Instant, T)>,
+}
+
+impl<T> Timeouts<T> {
+    fn new() -> Self {
+        Timeouts { values: Vec::new() }
+    }
+
+    fn push(&mut self, timeout: Instant, data_id: T) {
+        self.values.push((timeout, data_id));
+    }
+
+    fn purge(&mut self, now: &Instant) -> impl Iterator<Item = T> + '_ {
+        // The values are sorted by timeout.  Locate the index of the first non-expired one.
+        let split_index = match self
+            .values
+            .binary_search_by(|(timeout, _data_id)| timeout.cmp(now))
+        {
+            Ok(index) => index,
+            Err(index) => index,
+        };
+
+        // Drain and return the expired IDs.
+        self.values
+            .drain(..split_index)
+            .map(|(_timeout, data_id)| data_id)
+    }
+}
+
 #[derive(DataSize, Debug)]
 pub(crate) struct GossipTable<T> {
     /// Data IDs for which gossiping is still ongoing.
     current: HashMap<T, State>,
-    /// Data IDs for which gossiping is complete.  The map's values are the times after which the
-    /// relevant entries can be removed.
-    finished: HashMap<T, Instant>,
+    /// Data IDs for which gossiping is complete.
+    finished: HashSet<T>,
+    /// Timeouts for removal of items from the `finished` cache.
+    #[data_size(skip)]
+    finished_timeouts: Timeouts<T>,
     /// Data IDs for which gossiping has been paused (likely due to detecting that the data was not
     /// correct as per our current knowledge).  Such data could later be decided as still requiring
     /// to be gossiped, so we retain the `State` part here in order to resume gossiping.
-    paused: HashMap<T, (State, Instant)>,
+    paused: HashMap<T, State>,
+    /// Timeouts for removal of items from the `paused` cache.
+    #[data_size(skip)]
+    paused_timeouts: Timeouts<T>,
```

Comment:
We're storing every value twice now – once in `paused` and `finished` and once in `*_timeouts` collection. Instead of cloning whole structs, can we store IDs (in both collection) and refer to them by that? The easiest would be to use a hash of `T` (since we store them in `HashMap` and `HashSet` they already have to be `Hash`-able). Best would be to reuse `T`'s own "id" (because we know they will have it; it's either `Deploy` (which has `hash()`) or `Block` (also hash `hash()`)).

Reply:
The `T` here actually is the ID.  In the gossiper component `T` is e.g. `Deploy`, but the corresponding gossip table's `T` is `DeployHash` (https://github.com/CasperLabs/casper-node/blob/master/node/src/components/gossiper.rs#L102).

Reply:
Ah, right. Thanks. We did see a lot of CPU being beaten by hashing function when things were inserted/delete/looked up in the collections. Maybe it makes sense to cache the IDs of `Deploy`/`Block` and reuse them instead of hashing every time.

Reply:
Do we want to ticket it @Fraser999 ? i.e. replacing derived `Hash` impl for those types with a call to (cached) `hash()` method?

Reply:
Yup, I created https://casperlabs.atlassian.net/browse/NDRS-518.